don't replicate the deleted documents for WQ

ticoann opened the pull request:
It seems deleting document happens locally (WorkQueueManager and cronjob) so this seems not necessary and only creates the overhead. However, this needs further test. (Don't merge yet)

ticoann committed:
don't replicate the deleted documents

amaltaro commented:
Ok, getting back to this PR since couchdb is having a hard time to eat through the deleted docs during each couch restart.

If the local and cmsweb workqueue docs cleanup is performed always locally, then there should be no need to replicated these deleted docs.
For my record, the replication is started on the agent side following these directions:
 1) local/workqueue_inbox -> cmsweb/workqueue
 2) cmsweb/workqueue -> local/workqueue_inbox

amaltaro commented:
Ok, changes to the replication filter must be applied on both ends. If we apply it only to WMAgent, it still tries to replicate the deleted docs.

I setup a private VM with patched central services, also patched the WMAgent connected to it and I ran around 10 workflows yesterday and they all run properly, docs got deleted from both workqueue and workqueue_inbox in central couch, and the agent did not try to replicate the deleted docs.

The only issue I see is that there are some left over documents in the local couch (workqueue and workqueue_inbox), I picked 2 random and their statuses are Canceled or Done.

@ticoann do I understand it right that once the doc is pulled from central worqueue to workqueue_inbox, this doc gets duplicated into workqueue and it's used as a "sandbox", where all the updates are done afterwards? I see the same doc in both dbs, the only change is their revision (and a few parameters).

amaltaro commented:
Just an update (and questions). I ran around 30 wfs with this patch in and the only problem I saw is that the local workqueue_inbox docs were not deleted. I had to create this patch to clean those docs https://github.com/amaltaro/WMCore/commit/fdcf19d382cd32ace30750c915ddb2732ddf3ece

@ticoann I spent a few hours trying to understand it in details, but I failed :-( The only place workqueue (and _inbox) cleanup happens is here (right?)
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WorkQueue/WorkQueue.py#L820
these parents and elements twist my brain as well, I understand they have different meanings for global WQ and local WQ.

ticoann commented on src/python/WMCore/WorkQueue/WorkQueue.py:
```diff
@@ -816,8 +817,29 @@ def closeWork(self, *workflows):
+        reqNames = self.backend.getWorkflowNames()
+        reqNamesInbox = self.backend.flowNames(inboxFlag = True)
+        reqNames = list(set(reqNames).union(set(reqNamesInbox)))
+        requestsInfo = self.reqmgrSvc.getRequestByNames(reqNames)
```
TODO: This need to be more efficient by only returning RequestStatus instead of whole properties.

ticoann commented:
Alan, after discussing with you, I added the code to delete the WQE only when whole workflow is clean status. It is easier to do that instead of checking each element is synced between GQ and LQ. 
The problem which might cause by this is Element is not deleted until whole workflow is done which will have a lot bigger WQ db size. However this will give collect number for the % injected to LQ.
Please review, I will try to test mean while.

amaltaro commented on src/python/WMCore/WorkQueue/WorkQueue.py:
```diff
@@ -816,8 +817,29 @@ def closeWork(self, *workflows):
+        requestsInfo = self.reqmgrSvc.getRequestByNames(reqNames)
+        deleteRequests = []
+        for key, value in requestsInfo.items():
+            if ((value["RequestStatus"] == None) or \
```
How about we create a list of the end states and check the workflow status against that list?
Why don't you also test it against "completed", "closed-out"? I believe we do not need the wq
elements anymore at this point.

ticoann replied:
The problem is deleting WQ is parallel to state change. 
Workflow can be archived before WQ elements gets deleted. If that happens,WQ elements will never get deleted. Getting status of the workflow should include the archived workflow if we want to prevent that. But that will be too expensive. Getting workflow from WQ will limit the active workflows (~1000). I think that could be more efficient. (I need to check whether that call is ' POST' call though).

ticoann replied:
But I should add completed, closed-out, failed states as well to clean WQ as soon as possible. I will fix that.

amaltaro commented on src/python/WMCore/WorkQueue/WorkQueue.py:
```diff
@@ -816,8 +817,29 @@ def closeWork(self, *workflows):
+        deletes Workflow when workflow is in finished status
+        """
+        reqNames = self.backend.getWorkflowNames()
+        reqNamesInbox = self.backend.flowNames(inboxFlag = True)
```
typo

amaltaro commented on src/python/WMCore/WorkQueue/WorkQueueBackend.py:
```diff
@@ -524,3 +524,41 @@ def getWMBSInjectStatus(self, request = None):
+                           {'stale': "update_after", 'reduce' : True, 'group' : True})
+        return [x['key'] for x in data.get('rows', [])]
+    
+    def deletWQElementsByWorkflow(self, workflowNames):
```
typo :)

amaltaro commented on src/python/WMCore/WorkQueue/WorkQueueBackend.py:
```diff
@@ -524,3 +524,41 @@ def getWMBSInjectStatus(self, request = None):
+        options["reduce"] = False
+        
+        for couchdb in dbs:
+            result = couchdb.loadView("WorkQueue", "elementsByWorkflow", options, workflowNames)
```
Is the doc "id" the same as doc "key". If so, then we don't need to query this view again, since we get these results in getWorkflowNames()

ticoann replied:
That will be different except GQ inbox. Other workqueue element ids are UUID and key is workflow name.

ticoann commented:
Thanks Alan for the quick review. (Although you shouldn't work this late when you are sick :smile: ).  Could you check again (please tomorrow :smile: )?

amaltaro commented on src/python/WMCore/WorkQueue/WorkQueue.py:
```diff
@@ -895,6 +910,11 @@ def performQueueCleanupActions(self, skipWMBS = False):
                                                                             for x in finished_elements]),
                                                                  ', '.join(wf_to_cancel))
         self.backend.recordTaskActivity('housekeeping', msg)
```
Not related to your patch, but can you fix this typo "RequsetName" in line 902.

ticoann replied:
Thanks, good catch.

amaltaro commented on src/python/WMCore/WorkQueue/WorkQueue.py:
```diff
@@ -817,7 +826,26 @@ def closeWork(self, *workflows):
+        reqNames = self.backend.getWorkflowNames()
+        reqNamesInbox = self.backend.getWorkflowNames(inboxFlag = True)
+        reqNames = list(set(reqNames).union(set(reqNamesInbox)))
+        requestsInfo = self.requestDB.getRequestByNames(reqNames)
```
Don't we need to check whether it's local or global WQ (ReqMgr or RequestDB)?

ticoann replied:
Actually RequestDB is needed both global and local. I changed the comment.

amaltaro commented on src/python/WMCore/WorkQueue/WorkQueue.py:
```diff
@@ -204,6 +206,15 @@ def __init__(self, logger = None, dbi = None, **params):
+        
+        # used for global WQ
+        if self.params.get('ReqMgrServiceURL'):
+            self.reqmgrSvc = ReqMgr(self.params['ReqMgrServiceURL'])
```
Sorry for my stupid comment, but why we cannot use RequestDBReader for global WQ as well?
I always confuse these 2 :-(

amaltaro commented:
Besides my 2 short comments, this patch looks good.
I'll download and apply it to an agent, so feel free to update it if you need

ticoann commented:
Thanks Alan for the review. I changed what you pointed out and updated the patch. I ran the test in my vm. It seems working fine. (we need more test though)
Following deployment script change is needed for GQ
https://github.com/dmwm/deployment/commit/8773739861603aff6663bb76a56765e226e2babb

Do you want me to merge it for the test? If the test goes well I will make the 1.0.5 version of the patch since this need to be updated on agent side when production GQ is updated

ticoann commented:
If you want to run some test, reqmgr2-dev.cern.ch is set up and connected to vocms008

amaltaro commented:
Thanks Seangchan! Let me run some workflows in vocms008 first then. Have you already applied this patch to your VM and vocms008?

ticoann commented:
Yes, those machines are upto date. Thanks.
